### PR TITLE
Add workflow to manage stale PRs

### PR DESCRIPTION
**Pull request by ZehMatt**
The current rules for this setup, it runs every day at 1:30, I just copied the example time.
PRs with 14 days inactivity will get a comment and a stale-pr label, if no one responds or removes the label it will be closed after 14 more days.
Backtrace issues will be marked stale after 60 days with stale-backtrace label, if nothing is done it will be closed after 1 more day.

### CODE REVIEW

File: .github/workflows/stale.yml
Comment (James103) [inline .github/workflows/stale.yml:13]:
```suggestion
          stale-issue-message: 'This issue is stale because it has been open for 30 days with no activity. Remove stale label or comment or this will be closed in 5 days.'
          stale-pr-message: 'This PR is stale because it has been open for 45 days with no activity. Remove stale label or comment or this will be closed in 5 days.'
```

**Comment by CookiePLMonster** [comment]
I think this makes sense for backtrace issues, but this kind of workflow tends to be **very** frustrating for PRs and human made issues. Could this be limited to just backtrace?

**Comment by Broxzier** [comment]
We've discussed this a couple times in the past, and this has some cons to it. Some old issues are still valid nowadays, most likely not with high priority, but still things we should eventually get to. Same goes for some #good-first-issue issues that have yet to be picked up, or other manually created ones. Is it possible to only apply this to issues that are created by backtrace.io?

As for PRs: I think this is a good thing. After 45 days of no activity, it's clear that there is no active development and it will only bitrot more. We might as well decide to take over a PR by that time, to get it in a mergable state, or close it manually.

**Comment by ZehMatt** [comment]
We can actually add conditions to what it can and can't touch

**Comment by CookiePLMonster** [comment]
> As for PRs: I think this is a good thing. After 45 days of no activity, it's clear that there is no active development and it will only bitrot more. We might as well decide to take over a PR by that time, to get it in a mergable state, or close it manually.

I don't disagree, but in this case 45 days might be too harsh, especially if a PR is waiting for some dependency. Auto-closing might still be good but it should be some generous limit, like half a year + another month to bump. Fast enough to avoid clutter, slow enough not to be that annoying.

**Comment by trzejos** [comment]
Maybe we can add a blocked label or something that would prevent PRs and issues from being marked as stale

**Comment by ZehMatt** [comment]
> > As for PRs: I think this is a good thing. After 45 days of no activity, it's clear that there is no active development and it will only bitrot more. We might as well decide to take over a PR by that time, to get it in a mergable state, or close it manually.
> 
> I don't disagree, but in this case 45 days might be too harsh, especially if a PR is waiting for some dependency. Auto-closing might still be good but it should be some generous limit, like half a year + another month to bump. Fast enough to avoid clutter, slow enough not to be that annoying. 5 days is quick enough to catch a maintainer mid-holiday and have them unable to respond in a timely manner.

Keep in mind that it first comments on the PR and after 5 more days it closes it, I can make it 60 days for the first comment and maybe 30 days before it finally closes it, thats very generous I think.

**Comment by Basssiiie** [comment]
As mentioned on Discord; please don't just close issues randomly. 90% of the time they're "stale" because they haven't been picked up yet, not because they've been magically fixed and it's so freaking frustrating in other Github repositories. You're just hiding issues that are still present.

There's plenty of issues currently in the issue tracker that are already years old but are still present in the latest develop. No one will be able to fix them all within 30 or 45 days and lots of them are not labelled either. 

Also personally I like to spend decent time on reporting detailed issues with information and media, but fuck that if it just gets closed after 45 days. It wouldn't be worth my time.

**Comment by Gymnasiast** [comment]
> We've discussed this a couple times in the past, and this has some cons to it. Some old issues are still valid nowadays, most likely not with high priority, but still things we should eventually get to. Same goes for some #good-first-issue issues that have yet to be picked up, or other manually created ones. Is it possible to only apply this to issues that are created by backtrace.io?

> As for PRs: I think this is a good thing. After 45 days of no activity, it's clear that there is no active development and it will only bitrot more. We might as well decide to take over a PR by that time, to get it in a mergable state, or close it manually.

IIRC, we did decide not to autoclose anything, but to just have the bot comment/mark stuff. Not sure if you were implying that, just figured I’d clarify.

**Comment by ZehMatt** [comment]
> As mentioned on Discord; please don't just close issues randomly. 90% of the time they're "stale" because they haven't been picked up yet, not because they've been magically fixed and it's so freaking frustrating in other Github repositories. You're just hiding issues that are still present.
> 
> There's plenty of issues currently in the issue tracker that are already years old but are still present in the latest develop. No one will be able to fix them all within 30 or 45 days and lots of them are not labelled either.
> 
> Also personally I like to spend decent time on reporting detailed issues with information and media, but fuck that if it just gets closed after 45 days. It wouldn't be worth my time.
> 
> Edit: I'm only talking about human made issues. Automatically generated backtrace issues, feel free to close them automatically.

It is specifically designed in a way to first comment on the issue and if still left untouched it will close it, it will not just go on major closing spree without giving us 10 days to react to that, for issues where we don't care we'll simply let it close and for issues that are still valid we remove the label and the period will be reset to 60 days. I've already added a bunch of exclusions, this also gives us the opportunity to actually check if the issue still persists when the bot is going to comment.

**Comment by karst** [comment]
It might be an idea to add priority badges, or have the issue pinged every now and then so some could take a look at it. We must admit there's an issue with the amount of open issues, it makes it nearly impossible to find something, especially through the spam from backtrace. It also makes duplicate issue reports happen more often, the more issues are open. 
If nobody has complained about a certain issue, is it a big deal that it's closed? Then if someone complains about it again, it will have new attention to it and the chance is higher someone will actually fix it.

You could also mark automatic generated with the tag "Automatic" and have those go stale a lot quicker than other outstanding issues. Let's say auto generated would be 45 days, while other issues would go stale after half a year, or a year, unless someone pings it otherwise, or is marked with a prio badge?

Just brainstorming

**Comment by ZehMatt** [comment]
> It might be an idea to add priority badges, or have the issue pinged every now and then so some could take a look at it. We must admit there's an issue with the amount of open issues, it makes it nearly impossible to find something, especially through the spam from backtrace. It also makes duplicate issue reports happen more often, the more issues are open. If nobody has complained about a certain issue, is it a big deal that it's closed? Then if someone complains about it again, it will have new attention to it and the chance is higher someone will actually fix it.
> 
> You could also mark automatic generated with the tag "Automatic" and have those go stale a lot quicker than other outstanding issues. Let's say auto generated would be 45 days, while other issues would go stale after half a year, or a year, unless someone pings it otherwise, or is marked with a prio badge?
> 
> Just brainstorming

Its also not like the issues just vanish, they can be still visited and re-opened at any given time. Any activity on the issue will prevent the bot from even marking it as stale before closing it 10 days later.

**Comment by Basssiiie** [comment]
> It is specifically designed in a way to first comment on the issue and if still left untouched it will close it, it will not just go on major closing spree without giving us 10 days to react to that, for issues where we don't care we'll simply let it close and for issues that are still valid we remove the label and the period will be reset to 60 days. I've already added a bunch of exclusions, this also gives us the opportunity to actually check if the issue still persists when the bot is going to comment. Also simply commenting will prevent from closing it.

Will the issues that the bot is involved be monitored in any way? How and where will that happen? Or does the issue creator or people involved in reporting it have to bump it every 10 days?

> Its also not like the issues just vanish, they can be still visited and re-opened at any given time. Any activity on the issue will prevent the bot from even marking it as stale before closing it 10 days later.

They can but will you or any other developers actively monitor or visit those issues if there are any further replies? In my experience with other repositories, this happens rarely or never. It was mentioned that this could remove 'clutter', but if you'd want to monitor them regardless, you're just moving them to a bigger pile of clutter (the closed issues pile).

**Comment by ZehMatt** [comment]
> > It is specifically designed in a way to first comment on the issue and if still left untouched it will close it, it will not just go on major closing spree without giving us 10 days to react to that, for issues where we don't care we'll simply let it close and for issues that are still valid we remove the label and the period will be reset to 60 days. I've already added a bunch of exclusions, this also gives us the opportunity to actually check if the issue still persists when the bot is going to comment. Also simply commenting will prevent from closing it.
> 
> Will the issues that the bot is involved with be monitored in any way? How and where will that happen? Or does the issue creator or people involved in reporting it have to bump it every 10 days?
> 
> > Its also not like the issues just vanish, they can be still visited and re-opened at any given time. Any activity on the issue will prevent the bot from even marking it as stale before closing it 10 days later.
> 
> They can but will you or any other developers actively monitor or visit those issues if there are any further replies? In my experience with other repositories, this happens rarely or never. It was mentioned that this could remove 'clutter', but if you'd want to monitor them regardless, you're just moving them to a bigger pile of clutter (the closed issues pile).

If no one responds to the issue or PR then it will be closed in 10 days or whatever the amount of days we choose, if someone responds or removes the label then it will be quiet for another 60 days or whatever amount we set. If the reporter is no longer responding and no one else is interested in the issue we'll likely know that its stale, if you are subscribed to the repo you will get notifications for every comment made so most of us will actually be in the loop. This is rather a good way to keep relevant issues relevant and let the rest die out.

**Comment by ZehMatt** [comment]
As suggested by @IntelOrca it only does PR's, notification after 14 days and without response after another 14 days it will be closed.

File: .github/workflows/stale-prs.yml
Comment (Broxzier) [inline .github/workflows/stale-prs.yml:15]:
```suggestion
          stale-pr-message: 'This pull request is stale and will be closed in 14 days if no action is taken. To keep it open, remove the "stale" label or leave a comment.'
```

File: .github/workflows/stale-prs.yml
Comment (Broxzier) [inline .github/workflows/stale-prs.yml:16]:
```suggestion
          close-pr-message: 'This pull request was closed due to inactivity.'
```

File: .github/workflows/stale-backtrace-issues.yml
Comment (Broxzier) [inline .github/workflows/stale-backtrace-issues.yml:16]: "1 day" doesn't match the `days-before-issue-close` value of 7 that's set earlier.

**Review by Broxzier — CHANGES_REQUESTED** [review]

File: .github/workflows/stale-backtrace-issues.yml
Comment (James103) [reply .github/workflows/stale-backtrace-issues.yml:16]: This matches the other requested change. "in 1 day" can be changed to "tomorrow" if needed.
```suggestion
          stale-pr-message: 'This issue is stale because it has been open for 60 days with no activity and will be closed in 1 day. To keep it open, remove the "stale" label or leave a comment.'
```

Comment (ZehMatt) [reply .github/workflows/stale-backtrace-issues.yml]: It's meant to be 1 day, 61 days is more than enough time for those.

**Comment by Basssiiie** [comment]
One side note is that IIRC regular contributors cannot add or remove labels, so the "remove the stale label" suggestion as first suggestion is rarely applicable for the reader.

**Comment by Gymnasiast** [comment]
We have merged PRs that went undiscussed for a _lot_ longer than 14 days. It seems like an incredibly short time to me. 2 months + another for people to comment would seem far more reasonable to me.

**Comment by ZehMatt** [comment]
> One side note is that IIRC regular users cannot add or remove labels, so the "remove the stale label" suggestion as first suggestion is rarely applicable for the reader.

True, re-ordered that.

> We have merged PRs that went undiscussed for a _lot_ longer than 14 days. It seems like an incredibly short time to me (for autoclosing). 2 months + another for people to comment would seem far more reasonable to me.

Its 28 days before it will be closed, after 14 days it will comment and add the label if there is no response by either us or the author then after 14 days it will be closed, seems quite reasonable in my opinion, as long there is activity in any form it won't be closed.

**Comment by spacek531** [comment]
Will all subscribers be notified of the inactivity warning or just the author? On occasion I have PRs waiting for review which takes longer than 14 days. Will this system automatically poke team members for review?

**Comment by ZehMatt** [comment]
> Will all subscribers be notified of the inactivity warning or just the author? On occasion I have PRs waiting for review which takes longer than 14 days. Will this system automatically poke team members for review?

Its jus an ordinary comment, so either as author, participant or simply subscribing to all should show you a notification.

**Comment by Basssiiie** [comment]
I'd assume it's 14 days after the last comment, not 14 days after PR creation, right? Otherwise you'd have a decent amount of stale bot spam in your discussion.

File: .github/workflows/stale-backtrace-issues.yml
Comment (Broxzier) [inline .github/workflows/stale-backtrace-issues.yml:16]: Slightly more concise and straightforward rewording:

```suggestion
          stale-pr-message: 'This issue is stale and will be closed tomorrow if no action is taken. To keep it open, leave a comment or remove the "stale" label.'
```

File: .github/workflows/stale-backtrace-issues.yml
Comment (James103) [inline .github/workflows/stale-backtrace-issues.yml:1]:
```suggestion
name: 'Manage stale issues'
```

**Comment by Gymnasiast** [comment]
Often PRs get unnoticed by the team, or the team members are slow to react. Autoclosing PRs means that PRs like https://github.com/OpenRCT2/OpenRCT2/pull/16241 would have been autoclosed without any comment from a team member. That’s going to discourage outside contributors from contributing. Only last month (when I was still in the team) we decided not to autoclose anything for reasons like this.

**Comment by jensj12** [comment]
Due to the nature of open-source projects, handling of issues and PRs can be slow at times (there have been cases here where it was over 1 year before any activity happened on a now-merged PR, my apologies for that one).
PR's should only be closed for good reasons. While inactivity may sometimes hint at a good reason (abandoned, outdated), it is none in itself, as there can be other reasons for inactivity that do not require closing of the PR (dependencies, lack of time, forgotten). I consider looking though all open PRs (and to a lesser extend issues) every now and then part of maintaining a repository (which can be done with `is:pr is:open sort:updated-asc` and labels like `On hold` to avoid duplicate work). Adding a timer on all PRs only forces that task to be done regularly to not have PRs closed without a good reason. An automated warning is probably fine.

**Comment by ZehMatt** [comment]
> Due to the nature of open-source projects, handling of issues and PRs can be slow at times (there have been cases here where it was over 1 year before any activity happened on a now-merged PR, my apologies for that one). PR's should only be closed for good reasons. While inactivity may sometimes hint at a good reason (abandoned, outdated), it is none in itself, as there can be other reasons for inactivity that do not require closing of the PR (dependencies, lack of time, forgotten). I consider looking though all open PRs (and to a lesser extend issues) every now and then part of maintaining a repository (which can be done with `is:pr is:open sort:updated-asc` and labels like `On hold` to avoid duplicate work). Adding a timer on all PRs only forces that task to be done regularly to not have PRs closed without a good reason. An automated warning is probably fine.

We can add exceptions for PR's that have certain labels, in any case PR's won't be simply closed, the team can still remove the label or the author can simply respond that he needs more time and we can proceed with the appropriate actions.

**Review by Broxzier — APPROVED** [review]
The messages seem better now. Curious to see how well auto-closing PRs after 28 days of inactivity will work.

**Review by tupaschoal — APPROVED** [review]
It looks good to me with the current rules.

File: .github/workflows/stale-prs.yml
Comment (Broxzier) [inline .github/workflows/stale-prs.yml:15]: We might want to add instructions for the author to ping a developer if it's ready to be merged.

Comment (ZehMatt) [reply .github/workflows/stale-prs.yml:15]: How would you phrase it?

Comment (Broxzier) [reply .github/workflows/stale-prs.yml:15]: Something like:
```suggestion
          stale-pr-message: 'This pull request is stale and will be closed in 14 days if no action is taken. To keep it open, leave a comment or remove the "stale" label. If you\'re awating feedback from a developer, please send us a reminder (either here or on Discord).'
```

File: .github/workflows/stale-backtrace-issues.yml
Comment (janisozaur) [inline .github/workflows/stale-backtrace-issues.yml:16]:
```suggestion
          stale-pr-message: 'This issue is stale and will be closed tomorrow if no action is taken. To keep it open, leave a comment or remove the `stale-backtrace` label.'
```


Can anyone leave a comment or just the developers?

File: .github/workflows/stale-prs.yml
Comment (janisozaur) [inline .github/workflows/stale-prs.yml:15]:
```suggestion
          stale-pr-message: 'This pull request is stale and will be closed in 14 days if no action is taken. To keep it open, leave a comment or remove the `stale-pr` label. If you're awaiting feedback from a developer, please send us a reminder (either here or on Discord).'
```

Maybe linkify "discord" (with the link from readme)?

File: .github/workflows/stale-backtrace-issues.yml
Comment (ZehMatt) [reply .github/workflows/stale-backtrace-issues.yml:16]: Anyone, see the documentation about the action, its based on last update status of each issue/PR which comments update.

File: .github/workflows/stale-prs.yml
Comment (janisozaur) [inline .github/workflows/stale-prs.yml:15]:
```suggestion
          stale-pr-message: 'This pull request is stale and will be closed in 14 days if no action is taken. To keep it open, leave a comment or remove the `stale-pr` label. If you're awaiting feedback from a developer, please send us a reminder (either here or on [Discord](https://discord.gg/fsEwSWs)).'
```

Comment (janisozaur) [inline .github/workflows/stale-prs.yml:15]:
```suggestion
          stale-pr-message: 'This pull request is stale and will be closed in 14 days if no action is taken. To keep it open, leave a comment or remove the `stale-pr` label. If you''re awaiting feedback from a developer, please send us a reminder (either here or on [Discord](https://discord.gg/ZXZd8D8)).'
```

We use that link throughout the project, in readme, in-game and in release annoucments, let's keep this one only

Comment (ZehMatt) [reply .github/workflows/stale-prs.yml:15]: This is invalid syntax

Comment (janisozaur) [reply .github/workflows/stale-prs.yml:15]: I see

**Pull request merged**
